src/ntpmon.py

```diff
@@ -81,23 +82,29 @@ def main():
     if args.interval is None:
         args.interval = 60
 
-    if args.mode == 'telegraf':
+    if args.mode == 'telegraf' and not sys.stdout.isatty():
         (host, port) = args.connect.split(':')
         port = int(port)
         s = socket.socket()
         s.connect((host, port))
         sys.stdout = s.makefile(mode='w')
 
-    alerter = NTPAlerter(checks)
+    alerter = alert.NTPAlerter(checks)
+    implementation = None
     while True:
-        # run the checks
-        checkobjs = ntpchecks(checks, debug=False)
+        # cache implementation for the lifetime of ntpmon
+        if not implementation:
+            implementation = process.detect_implementation()
```

Comment:
Should this be warning if there is no detectable implementation? At the moment it could run forever, emitting no alerts if the service wasn't setup (arguably correct, or arguably a critical failure)

Reply:
From memory, this arose because the ntp charm might be installed before the telegraf charm, but the daemon might not be started.  So it's possible during deploys for no running NTP implementation to be available on ntpmon startup, but for one to appear shortly afterwards.  Startup is gated on one or other implementation's config file existing; if the config file is removed, the when_file_changed() handlers should notice and stop the service.  The cost of a ntpmon running forever and not detecting an implementation seems less than a constantly restarting systemd service.